protocol/messenger_store_node_request_manager.go

```diff
@@ -0,0 +1,327 @@
+	return &storeNodeRequest{
+		manager:       m,
+		subscriptions: make([]*communitySubscriptionChannel, 0),
+	}
+}
+
+// getFilter checks if a filter for a given community is already created and creates one of not found.
+// Returns the found/created filter, a flag if the filter was created by the function and an error.
+func (m *StoreNodeRequestManager) getFilter(c communities.CommunityShard) (*transport.Filter, bool, error) {
+	// First check if such filter already exists.
+	filter := m.messenger.transport.FilterByChatID(c.CommunityID)
+	if filter != nil {
+		//we don't remember filter id associated with community because it was already installed
+		return filter, false, nil
+	}
+
+	// If filter wasn't installed we create it and
+	// remember for uninstalling after response is received
+	filters, err := m.messenger.transport.InitPublicFilters([]transport.FiltersToInitialize{{
+		ChatID:      c.CommunityID,
+		PubsubTopic: c.Shard.PubsubTopic(),
+	}})
+
+	if err != nil {
+		m.logger.Error("can't install filter for community", zap.Error(err))
+		return nil, true, err
+	}
+
+	if len(filters) != 1 {
+		m.logger.Error("Unexpected number of filters created")
+		return nil, true, fmt.Errorf("unexepcted number of filters created")
+	}
+
+	return filters[0], true, nil
+}
+
+// forgetFilter uninstalls the given filter
+func (m *StoreNodeRequestManager) forgetFilter(filter *transport.Filter) {
+	err := m.messenger.transport.RemoveFilters([]*transport.Filter{filter})
+	if err != nil {
+		m.logger.Warn("failed to remove filter", zap.Error(err))
+	}
+}
+
+// storeNodeRequest represents a single store node batch request.
+// For a valid storeNodeRequest to be performed, the user must set all the struct fields and call start method.
+type storeNodeRequest struct {
+	// request parameters
+	pubsubTopic  string
+	contentTopic types.TopicType
+
+	// request corresponding metadata to be used in finalize
+	filterToForget *transport.Filter
+	communityID    string
+
+	// internal fields
+	manager       *StoreNodeRequestManager
+	subscriptions []*communitySubscriptionChannel
+	result        fetchCommunityResult
+}
+
+// fetchCommunityResult contains result of a single storeNodeRequest
+// If any error occurs during the request, err field will be set.
+// If a community was successfully fetched, community field will contain the fetched information.
+// If a community wasn't found in store node, then a community will be set to `nil`.
+// stats will contain information about the performed request that might be useful for testing.
+type fetchCommunityResult struct {
+	err       error
+	community *communities.Community
+	stats     FetchCommunityStats
+}
+
+type communitySubscriptionChannel = chan fetchCommunityResult
+
+func (r *storeNodeRequest) subscribe() *communitySubscriptionChannel {
+	channel := make(communitySubscriptionChannel, 100)
+	r.subscriptions = append(r.subscriptions, &channel)
+	return &channel
+}
```

Comment:
You can return the channel directly. No need to return a reference to it as channels already behave like references. (This will require changing `r.subscriptions` to not store channel references)

```suggestion
func (r *storeNodeRequest) subscribe() communitySubscriptionChannel {
	channel := make(communitySubscriptionChannel, 100)
	r.subscriptions = append(r.subscriptions, channel)
	return channel
}
```